Adds RigidBodyTree::get_num_actuators() and Generalizes GravityCompensator To Support Underactuated Models

liangfok opened the pull request:
This builds upon #4083. It adds `RigidBodyTree::get_num_actuators()` and generalizes `GravityCompensator` to support underactuated robots.

<!-- Reviewable:start -->

---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/4099)
<!-- Reviewable:end -->

liangfok commented:
+@naveenoid for feature review since you were the original author of `GravityCompensator`.
+@ggould-tri for platform review per rotation schedule.

Thanks!

-@liangfok

---

Review status: 0 of 5 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-:-KWstBKGAbm_CpvrLkuP:bn6mms8)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: 0 of 9 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[drake/systems/controllers/test/gravity_compensator_test.cc, line 72 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-KWt0bivoeZXe16vXiTa:-KWt0bivoeZXe16vXiTb:b-z5wqxm) ([raw file](https://github.com/robotlocomotion/drake/blob/aafa1b395a5507b57060956d836cd7aa610a4fb7/drake/systems/controllers/test/gravity_compensator_test.cc#L72)):*
> ```cpp
>   }
> 
>   std::unique_ptr<RigidBodyTreed> tree_;
> ```

BTW I think this looked better as `RigidBodyTree<double>` for consistency with the next four items.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4099)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 0 of 9 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/controllers/test/gravity_compensator_test.cc, line 72 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-KWt0bivoeZXe16vXiTa:-KWtHjVQwunbKgG36Wmq:b-uq91r4) ([raw file](https://github.com/robotlocomotion/drake/blob/aafa1b395a5507b57060956d836cd7aa610a4fb7/drake/systems/controllers/test/gravity_compensator_test.cc#L72)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW I think this looked better as `RigidBodyTree<double>` for consistency with the next four items.
</details>
Done. For additional consistency, I switched from using `RigidBodyTreed` to `RigidBodyTree<double>` throughout this file.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4099)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I'm going to apply the "do no review" tag until after I cleanup the git commit history.

liangfok commented:
This PR passed all CI checks and is thus ready for review.

sherm1 commented:
-@ggould-tri +@sherm1 for platform review. I have a few non-blocking comments, PTAL. Then: <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>.

---

Reviewed 4 of 5 files at r1, 4 of 4 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/controllers/gravity_compensator.cc, line 52 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-KWy7WK_DRBF3g9y-gZl:-KWy7WK_DRBF3g9y-gZm:b9h8erw) ([raw file](https://github.com/robotlocomotion/drake/blob/849961874e810bcb94acb586a458e576b23084d7/drake/systems/controllers/gravity_compensator.cc#L52)):*
> ```cpp
>   }
>   DRAKE_ASSERT(actuated_g.size() ==
>                System<T>::GetMutableOutputVector(output, 0).size());
> ```

BTW, an assert shouldn't ask for mutable access which might trigger computation.

---

*[drake/systems/controllers/test/gravity_compensator_test.cc, line 41 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-KWy82hQGeXYZh45kPUz:-KWy82hQGeXYZh45kPV-:b9qq7zk) ([raw file](https://github.com/robotlocomotion/drake/blob/849961874e810bcb94acb586a458e576b23084d7/drake/systems/controllers/test/gravity_compensator_test.cc#L41)):*
> ```cpp
>   // need to extract from `g` the torque / force commands corresponding to the
>   // actuators that are used.
>   Eigen::VectorXd actuated_g(rigid_body_tree.get_num_actuators());
> ```

BTW, this code duplicated from gravity_compensator -- can you avoid that?

---

*[drake/systems/controllers/test/gravity_compensator_test.cc, line 106 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-KWy8gSzP0vvbrDxnHoK:-KWy8gT-xL9GUG3FjjDP:b-5kuobt) ([raw file](https://github.com/robotlocomotion/drake/blob/849961874e810bcb94acb586a458e576b23084d7/drake/systems/controllers/test/gravity_compensator_test.cc#L106)):*
> ```cpp
>   // to this unit test's instance of `GravityCompensatorTest`. We quarantine
>   // this local variable to prevent downstream code from seg faulting by trying
>   // to access `tree` after ownership is transferred.
> ```

BTW, I'm in favor of quarantining local variables but I don't understand the need for this comment. The code below shouldn't be accessing the `tree` local at all even if it were still in scope -- *that* would prevent seg faulting!

---

*[drake/systems/controllers/test/gravity_compensator_test.cc, line 142 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-KWy9F5uW1c0I2XAI9yo:-KWy9F5uW1c0I2XAI9yp:b2hy4cr) ([raw file](https://github.com/robotlocomotion/drake/blob/849961874e810bcb94acb586a458e576b23084d7/drake/systems/controllers/test/gravity_compensator_test.cc#L142)):*
> ```cpp
>   // to this unit test's instance of `GravityCompensatorTest`. We quarantine
>   // this local variable to prevent downstream code from seg faulting by trying
>   // to access `tree` after ownership is transferred.
> ```

BTW, nuke most of this comment?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-:-KWy9x5P9CN-JLxX1P3c:bti9ucs)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
Finished the first pass. Thanks an PTAL.

---

Reviewed 3 of 5 files at r1, 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/systems/controllers/gravity_compensator.cc, line 49 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-KX5c2E1lhIHrgzVvi4T:-KX5c2E1lhIHrgzVvi4U:b-qinsqc) ([raw file](https://github.com/robotlocomotion/drake/blob/849961874e810bcb94acb586a458e576b23084d7/drake/systems/controllers/gravity_compensator.cc#L49)):*
> ```cpp
>     int index_in_g =
>         rigid_body_tree_.actuators.at(i).body_->get_position_start_index();
>     actuated_g[i] = g[index_in_g];
> ```

It is not clear to me how this bit of logic works. In case of a general underactuated system, the actuation input is mapped to the generalised forces by an input matrix (in the static case, G(q) = B(q)u ) and the inversion of the structure represented by B(q) is needed to compensate for gravity. I accept that for the fully actuated case, this bit of code will still function but Its not clear how it would be able to achieve gravity compensation for systems such as acrobot or quadrotor. 

---

*[drake/systems/controllers/test/gravity_compensator_test.cc, line 28 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-KX5fQu_pTeoXh20zBeK:-KX5fQu_pTeoXh20zBeL:b16rfle) ([raw file](https://github.com/robotlocomotion/drake/blob/849961874e810bcb94acb586a458e576b23084d7/drake/systems/controllers/test/gravity_compensator_test.cc#L28)):*
> ```cpp
> namespace {
> 
> VectorXd ComputeIiwaGravityTorque(const RigidBodyTree<double>& rigid_body_tree,
> ```

This method might need renaming to ComputeRobotGravityTorque since it is being used for the 4 bar linkage example below as well.

---

*[drake/systems/controllers/test/gravity_compensator_test.cc, line 50 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-KX5cbnpRXdQptzEroBP:-KX5cbnpRXdQptzEroBQ:b-wzh697) ([raw file](https://github.com/robotlocomotion/drake/blob/849961874e810bcb94acb586a458e576b23084d7/drake/systems/controllers/test/gravity_compensator_test.cc#L50)):*
> ```cpp
>                      .get_num_positions() == 1);
>     int index_in_g =
>         rigid_body_tree.actuators.at(i).body_->get_position_start_index();
> ```

BTW: as an alternative to repeating the computation inside gravity compensation, Im wondering if its better to compute the gravity torque, feed it into the system being controlled and then demonstrate that the resulting motion is of 0 acceleration and 0 velocity instead. This might overcome hidden problems in repeating the controller logic as a verification. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-:-KX5g4-sMXOGHoQ04fT0:b-xcj9k)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/systems/controllers/gravity_compensator.cc, line 49 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-KX5c2E1lhIHrgzVvi4T:-KX9PLjiTKL9D90WwiVG:b-qk7m1h) ([raw file](https://github.com/robotlocomotion/drake/blob/849961874e810bcb94acb586a458e576b23084d7/drake/systems/controllers/gravity_compensator.cc#L49)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary>
> It is not clear to me how this bit of logic works. In case of a general underactuated system, the actuation input is mapped to the generalised forces by an input matrix (in the static case, G(q) = B(q)u ) and the inversion of the structure represented by B(q) is needed to compensate for gravity. I accept that for the fully actuated case, this bit of code will still function but Its not clear how it would be able to achieve gravity compensation for systems such as acrobot or quadrotor. 
</details>
Perhaps I'm not understanding the semantics of the method call that computes `g`:


```
Eigen::VectorXd g = rigid_body_tree_.dynamicsBiasTerm(
    cache, f_ext, false /* include velocity terms */);
```

I'm observing that it is returning a gravity-compensating force / torque for every `RigidBody` in the tree, not just those that are connected to actuated joints. Is this correct? If so, how do I map `g` to the desired output vector, which is a vector for only the actuated joints?

---

*[drake/systems/controllers/test/gravity_compensator_test.cc, line 106 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-KWy8gSzP0vvbrDxnHoK:-KWyHmsfjXIFAOAS6qB2:b-zew7f9) ([raw file](https://github.com/robotlocomotion/drake/blob/849961874e810bcb94acb586a458e576b23084d7/drake/systems/controllers/test/gravity_compensator_test.cc#L106)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, I'm in favor of quarantining local variables but I don't understand the need for this comment. The code below shouldn't be accessing the `tree` local at all even if it were still in scope -- *that* would prevent seg faulting!
</details>
Yes, the subsequent code should not (and does not) access `tree`, but it does access `tree_`. I've removed the comment but kept the scoping curly brace just to reinforce the face that `tree_` is the only choice.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4099)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/systems/controllers/gravity_compensator.cc, line 49 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-KX5c2E1lhIHrgzVvi4T:-KXAIJHqeP8R2brNL-7R:blv9upt) ([raw file](https://github.com/robotlocomotion/drake/blob/849961874e810bcb94acb586a458e576b23084d7/drake/systems/controllers/gravity_compensator.cc#L49)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Perhaps I'm not understanding the semantics of the method call that computes `g`:
> 
> 
> ```
> Eigen::VectorXd g = rigid_body_tree_.dynamicsBiasTerm(
>     cache, f_ext, false /* include velocity terms */);
> ```
> 
> I'm observing that it is returning a gravity-compensating force / torque for every `RigidBody` in the tree, not just those that are connected to actuated joints. Is this correct? If so, how do I map `g` to the desired output vector, which is a vector for only the actuated joints?
</details>
From the documentation of dynamic bias term :  
\*Compute
  * \f[
  *  H(q) \dot{v} + C(q, v, f_\text{ext})
  * \f]
  * that is, the left hand side of the manipulator equations
  *  \f[
  *  H(q) \dot{v} + C(q, v, f_\text{ext}) = B(q) u
  * \f]
  *
  * Note that the 'dynamics bias term' \f$ C(q, v, f_\text{ext}) \f$ can be
  *computed by simply setting \f$ \dot{v} = 0\f$.
*\

i.e. you are computing the left side of the equation. Whereas, mapping to the actual actuator inputs requires the inversion of the structure represented by B(q). The solution you propose in the code above will only be applicable to simple underactuated systems (underactuated kinematic chains) where B(q) is composed of 1s and 0s and that too having a unique 1 in every column. That is why the inversion will uniquely map 1 DoF to 1 actuator for the fully actuated case. This wont work for systems such as quadrotor where B(q) must be computed and then inverted (you can only compensate gravity to .  Now AFAIU this is being done implicitly in the algorithms behind the scenes in RigidBodyTree but the API for B(q) needs exposing. Until then this kind of underactuated gravity compensation is only functional for underactuated kinematic chains and this must be noted. 

Tagging @russtedrake and @amcastro-tri for their comments on my observations.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4099)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/systems/controllers/gravity_compensator.cc, line 49 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-KX5c2E1lhIHrgzVvi4T:-KXB0vjt2O0GupDGf2Js:b-8fyqts) ([raw file](https://github.com/robotlocomotion/drake/blob/849961874e810bcb94acb586a458e576b23084d7/drake/systems/controllers/gravity_compensator.cc#L49)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary>
> From the documentation of dynamic bias term :  
> \*Compute
>   * \f[
>   *  H(q) \dot{v} + C(q, v, f_\text{ext})
>   * \f]
>   * that is, the left hand side of the manipulator equations
>   *  \f[
>   *  H(q) \dot{v} + C(q, v, f_\text{ext}) = B(q) u
>   * \f]
>   *
>   * Note that the 'dynamics bias term' \f$ C(q, v, f_\text{ext}) \f$ can be
>   *computed by simply setting \f$ \dot{v} = 0\f$.
> *\
> 
> i.e. you are computing the left side of the equation. Whereas, mapping to the actual actuator inputs requires the inversion of the structure represented by B(q). The solution you propose in the code above will only be applicable to simple underactuated systems (underactuated kinematic chains) where B(q) is composed of 1s and 0s and that too having a unique 1 in every column. That is why the inversion will uniquely map 1 DoF to 1 actuator for the fully actuated case. This wont work for systems such as quadrotor where B(q) must be computed and then inverted (you can only compensate gravity to .  Now AFAIU this is being done implicitly in the algorithms behind the scenes in RigidBodyTree but the API for B(q) needs exposing. Until then this kind of underactuated gravity compensation is only functional for underactuated kinematic chains and this must be noted. 
> 
> Tagging @russtedrake and @amcastro-tri for their comments on my observations.
</details>
Actually on second thought : What is the meaning of gravity compensation for an underactuated system? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4099)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Reviewed 2 of 5 files at r1, 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/systems/controllers/gravity_compensator.cc, line 49 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-KX5c2E1lhIHrgzVvi4T:-KXBnxe3Jlv2CzdsY-1k:bticqb8) ([raw file](https://github.com/robotlocomotion/drake/blob/849961874e810bcb94acb586a458e576b23084d7/drake/systems/controllers/gravity_compensator.cc#L49)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary>
> Actually on second thought : What is the meaning of gravity compensation for an underactuated system? 
</details>
To be honest I don't see what this code accomplishes. I think @naveenoid question is the one we have to pose first, what does gravity compensation mean for an under-actuated system?
I just went through the following thought experiment: Imagine a robot with two links and and two revolute joints (essentially a double pendulum or a Kuka with just two joints). Now think that the only actuated joint is the one connecting the two links (an acrobot!). 
So what the code above does is to return the torque to be applied on this actuator only. With zero torque at the base of the acrobot clearly the robot will start falling down. So no "gravity compensation" achieved at all.
@siyuanfeng-tri, what do you do for systems like Valkyrie?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4099)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/systems/controllers/gravity_compensator.cc, line 49 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-KX5c2E1lhIHrgzVvi4T:-KXByBOQE9AYFYd2pTpb:bsjz7g9) ([raw file](https://github.com/robotlocomotion/drake/blob/849961874e810bcb94acb586a458e576b23084d7/drake/systems/controllers/gravity_compensator.cc#L49)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> To be honest I don't see what this code accomplishes. I think @naveenoid question is the one we have to pose first, what does gravity compensation mean for an under-actuated system?
> I just went through the following thought experiment: Imagine a robot with two links and and two revolute joints (essentially a double pendulum or a Kuka with just two joints). Now think that the only actuated joint is the one connecting the two links (an acrobot!). 
> So what the code above does is to return the torque to be applied on this actuator only. With zero torque at the base of the acrobot clearly the robot will start falling down. So no "gravity compensation" achieved at all.
> @siyuanfeng-tri, what do you do for systems like Valkyrie?
> 
</details>
@naveenoid and @amcastro-tri for an under-actuated system, we can still do gravity compensation, the idea is just to use the actuator torque and external force to cancel the gravity term in the dynamics function.
```
H(q)*v_dot + C(q,v)*v + g(q) = B*u + J'*f_ext`
```
Notice here I expand your `C(q, v,f_ext)`. `C(q,v)*v` is the Coriolos term, and `g(q)` is the gravity term that only depends on robot posture.

When we do gravity compensation, we want `vdot, v` being both zero. Thus we want to find the actuator torque `u` and external force `lambda`, such that
```
g(q) = B*u + J'*f_ext
```
This is again a linear constraint on `u` and `f_ext`.

We have done gravity compensation for Atlas robot. I cannot find a video for our Atlas robot, but here is a video footage for another robot under gravity compensation
https://www.youtube.com/watch?v=18-xpWZXZAo

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4099)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/systems/controllers/gravity_compensator.cc, line 49 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-KX5c2E1lhIHrgzVvi4T:-KXC-dfpHiwmZvxXG5BQ:bpb7knj) ([raw file](https://github.com/robotlocomotion/drake/blob/849961874e810bcb94acb586a458e576b23084d7/drake/systems/controllers/gravity_compensator.cc#L49)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary>
> @naveenoid and @amcastro-tri for an under-actuated system, we can still do gravity compensation, the idea is just to use the actuator torque and external force to cancel the gravity term in the dynamics function.
> ```
> H(q)*v_dot + C(q,v)*v + g(q) = B*u + J'*f_ext`
> ```
> Notice here I expand your `C(q, v,f_ext)`. `C(q,v)*v` is the Coriolos term, and `g(q)` is the gravity term that only depends on robot posture.
> 
> When we do gravity compensation, we want `vdot, v` being both zero. Thus we want to find the actuator torque `u` and external force `lambda`, such that
> ```
> g(q) = B*u + J'*f_ext
> ```
> This is again a linear constraint on `u` and `f_ext`.
> 
> We have done gravity compensation for Atlas robot. I cannot find a video for our Atlas robot, but here is a video footage for another robot under gravity compensation
> https://www.youtube.com/watch?v=18-xpWZXZAo
</details>
@amcastro-tri @naveenoid @liangfok when we did the gravity compensation for the Atlas robot, we did not treat the external force `f_ext` as a known quantity (where does the external force measurement coming from? The force-torque sensor output is a DELAYED measurement of the actual external force, while the CURRENT external force is an output that directly depends on the actuator torque, thus you cannot use it for feedback control. Making your control torque `u` depending on CURRENT external force would cause a direct feedthrough problem). Instead, we treat the external force as an unknown decision variable, as we do for walking.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4099)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/systems/controllers/gravity_compensator.cc, line 49 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-KX5c2E1lhIHrgzVvi4T:-KXC74SQKoyj3Sdy1ixP:byke29e) ([raw file](https://github.com/robotlocomotion/drake/blob/849961874e810bcb94acb586a458e576b23084d7/drake/systems/controllers/gravity_compensator.cc#L49)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary>
> @amcastro-tri @naveenoid @liangfok when we did the gravity compensation for the Atlas robot, we did not treat the external force `f_ext` as a known quantity (where does the external force measurement coming from? The force-torque sensor output is a DELAYED measurement of the actual external force, while the CURRENT external force is an output that directly depends on the actuator torque, thus you cannot use it for feedback control. Making your control torque `u` depending on CURRENT external force would cause a direct feedthrough problem). Instead, we treat the external force as an unknown decision variable, as we do for walking.
</details>
Thanks for this explanation. Going back to the original objective of this PR, I was under the impression that this system should output `u`. Was I mistaken? Should the output of this system instead be `B*u`, which equals `g(q)` when `v = 0`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4099)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/systems/controllers/gravity_compensator.cc, line 49 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-KX5c2E1lhIHrgzVvi4T:-KXC8wY2l0qWfcYZ6K0-:bvvwe2z) ([raw file](https://github.com/robotlocomotion/drake/blob/849961874e810bcb94acb586a458e576b23084d7/drake/systems/controllers/gravity_compensator.cc#L49)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Thanks for this explanation. Going back to the original objective of this PR, I was under the impression that this system should output `u`. Was I mistaken? Should the output of this system instead be `B*u`, which equals `g(q)` when `v = 0`?
</details>
@liangfok I think you are right that we should output `u`. I think one of the confusion here is that you are supposing `B.transpose() = B.inverse()`, namely `B` is just a permutation matrix, with only one `1` in each column/row, and zero otherwise, like
```
B = [0 0 1]
    [1 0 0]
    [0 1 0]
```
This is not true in general.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4099)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/systems/controllers/gravity_compensator.cc, line 49 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-KX5c2E1lhIHrgzVvi4T:-KXCCiaDCewmRPie8NI6:b-1802ne) ([raw file](https://github.com/robotlocomotion/drake/blob/849961874e810bcb94acb586a458e576b23084d7/drake/systems/controllers/gravity_compensator.cc#L49)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary>
> @liangfok I think you are right that we should output `u`. I think one of the confusion here is that you are supposing `B.transpose() = B.inverse()`, namely `B` is just a permutation matrix, with only one `1` in each column/row, and zero otherwise, like
> ```
> B = [0 0 1]
>     [1 0 0]
>     [0 1 0]
> ```
> This is not true in general.
</details>
OK thanks again. I will update this PR to output B<sup>-1</sup> g(q).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4099)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/systems/controllers/gravity_compensator.cc, line 49 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-KX5c2E1lhIHrgzVvi4T:-KXCBRG86iNBAB3PUmkf:bxyzws3) ([raw file](https://github.com/robotlocomotion/drake/blob/849961874e810bcb94acb586a458e576b23084d7/drake/systems/controllers/gravity_compensator.cc#L49)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> OK thanks again. I will update this PR to output B<sup>-1</sup> g(q).
</details>
Sorry for being late to the party, I was at a seminar talk. I think the problem is not well defined for an under-actuated / no external force case.
I personally thing this should have the strictest assumptions to handle the simplest case, where every joint is actuated and decoupled, you achieve 0 (or a desired acceleration), there are not external forces / loads. And you throw exceptions if those are not met. 
For a more generalized block, we should do the QP inverse dynamics things to handle underaction, external loads, because fundamentally, they problem is ill defined, and you have to make a decision as how to comprise among different things, and that's where optimization comes in.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4099)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/systems/controllers/gravity_compensator.cc, line 49 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-KX5c2E1lhIHrgzVvi4T:-KXCDf348ZOpWK0WgyhE:b-r4wars) ([raw file](https://github.com/robotlocomotion/drake/blob/849961874e810bcb94acb586a458e576b23084d7/drake/systems/controllers/gravity_compensator.cc#L49)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary>
> Sorry for being late to the party, I was at a seminar talk. I think the problem is not well defined for an under-actuated / no external force case.
> I personally thing this should have the strictest assumptions to handle the simplest case, where every joint is actuated and decoupled, you achieve 0 (or a desired acceleration), there are not external forces / loads. And you throw exceptions if those are not met. 
> For a more generalized block, we should do the QP inverse dynamics things to handle underaction, external loads, because fundamentally, they problem is ill defined, and you have to make a decision as how to comprise among different things, and that's where optimization comes in.
> 
</details>
I agree with @siyuanfeng-tri , outputing `B.inverse()*g(q)` does not solve the problem here, since you are treating the external force as known quantity. We should do a QP inverse dynamics to solve the equation I mentioned above `g(q) = B*u + J'*f_ext`, to find out both the control torque `u` and the external force `f_ext`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4099)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/systems/controllers/gravity_compensator.cc, line 49 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-KX5c2E1lhIHrgzVvi4T:-KXCDsb-SBE1sVaJfuLC:bxgugc4) ([raw file](https://github.com/robotlocomotion/drake/blob/849961874e810bcb94acb586a458e576b23084d7/drake/systems/controllers/gravity_compensator.cc#L49)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary>
> I agree with @siyuanfeng-tri , outputing `B.inverse()*g(q)` does not solve the problem here, since you are treating the external force as known quantity. We should do a QP inverse dynamics to solve the equation I mentioned above `g(q) = B*u + J'*f_ext`, to find out both the control torque `u` and the external force `f_ext`.
</details>
Liang, I wouldn't go change it to B inverse at this point. I would stick to using B transpose, and if the system doesn't satisfy the assumptions for B transpose, I would throw an exceptions. Because handling 2 actuators to 3 dof torque is generally not quite straight forward. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4099)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/systems/controllers/gravity_compensator.cc, line 49 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-KX5c2E1lhIHrgzVvi4T:-KXCERRh5id8fi3X4woS:b-8jmg9g) ([raw file](https://github.com/robotlocomotion/drake/blob/849961874e810bcb94acb586a458e576b23084d7/drake/systems/controllers/gravity_compensator.cc#L49)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary>
> Liang, I wouldn't go change it to B inverse at this point. I would stick to using B transpose, and if the system doesn't satisfy the assumptions for B transpose, I would throw an exceptions. Because handling 2 actuators to 3 dof torque is generally not quite straight forward. 
</details>
Let's just keep it simple here, state the assumptions, and roll over and die if those are not met. This code is not supposed to handle those situations. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4099)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/systems/controllers/gravity_compensator.cc, line 49 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-KX5c2E1lhIHrgzVvi4T:-KXCEe84IOPLtCUZxVbl:beqj0i6) ([raw file](https://github.com/robotlocomotion/drake/blob/849961874e810bcb94acb586a458e576b23084d7/drake/systems/controllers/gravity_compensator.cc#L49)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary>
> Let's just keep it simple here, state the assumptions, and roll over and die if those are not met. This code is not supposed to handle those situations. 
</details>
I agree with @siyuanfeng-tri , we should not try to handle the under-actuated case here. Instead, the gravity compensation for underactuated system should be handled by a QP controller, in a separate file.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4099)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/systems/controllers/gravity_compensator.cc, line 49 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-KX5c2E1lhIHrgzVvi4T:-KXCRuxAYE4gnxiWHaIY:b-8e2ube) ([raw file](https://github.com/robotlocomotion/drake/blob/849961874e810bcb94acb586a458e576b23084d7/drake/systems/controllers/gravity_compensator.cc#L49)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary>
> I agree with @siyuanfeng-tri , we should not try to handle the under-actuated case here. Instead, the gravity compensation for underactuated system should be handled by a QP controller, in a separate file.
</details>
Thanks for all of the comments. After a F2F discussion with @amcastro-tri and @siyuanfeng-tri I've decided to modify this PR to simply throw an exception if it is given a model that is not fully actuated. We will handle underactuated models using different techniques on a case-by-case basis.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4099)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/systems/controllers/gravity_compensator.cc, line 49 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-KX5c2E1lhIHrgzVvi4T:-KXCd_9_d2ia1NWt6H0z:b8midk4) ([raw file](https://github.com/robotlocomotion/drake/blob/849961874e810bcb94acb586a458e576b23084d7/drake/systems/controllers/gravity_compensator.cc#L49)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Thanks for all of the comments. After a F2F discussion with @amcastro-tri and @siyuanfeng-tri I've decided to modify this PR to simply throw an exception if it is given a model that is not fully actuated. We will handle underactuated models using different techniques on a case-by-case basis.
</details>
Im ok with that conclusion! Thanks a lot @amcastro-tri @hongkai-dai @siyuanfeng-tri 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4099)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Reviewed 1 of 3 files at r4.
Review status: 8 of 10 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[drake/systems/controllers/gravity_compensator.cc, line 21 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-KXD8B76-0K7jS3S57MZ:-KXD8B76-0K7jS3S57M_:b4jn9yw) ([raw file](https://github.com/robotlocomotion/drake/blob/5dee5c4eb01760f728c7cd8cdf415616a676906e/drake/systems/controllers/gravity_compensator.cc#L21)):*
> ```cpp
>         << "  - size of gravity vector: " << tree.get_num_positions() << "\n"
>         << "  - number of actuators: " << tree.get_num_actuators();
>     throw std::runtime_error(msg.str());
> ```

` DRAKE_ABORT_MSG` for printing messages consistently. I think at this point (with all the utilities in `drake_assert.h`) we should be avoiding explicit calls to `throw`?

---

*[drake/systems/controllers/test/gravity_compensator_test.cc, line 48 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-KXD7dTh0MWRuPoAPVIN:-KXD7dTh0MWRuPoAPVIO:bt50a9z) ([raw file](https://github.com/robotlocomotion/drake/blob/5dee5c4eb01760f728c7cd8cdf415616a676906e/drake/systems/controllers/test/gravity_compensator_test.cc#L48)):*
> ```cpp
>   GravityCompensatorTest() {}
> 
>   void SetUp() override {}
> ```

do we need the empty `SetUp()`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4099)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I addressed all reviewer comments. PTAL. Thanks.

---

Review status: 7 of 10 files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/systems/controllers/gravity_compensator.cc, line 49 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-KX5c2E1lhIHrgzVvi4T:-KXClLR2SbubFHhiDd5t:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/849961874e810bcb94acb586a458e576b23084d7/drake/systems/controllers/gravity_compensator.cc#L49)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary>
> Im ok with that conclusion! Thanks a lot @amcastro-tri @hongkai-dai @siyuanfeng-tri 
</details>
Done.

---

*[drake/systems/controllers/gravity_compensator.cc, line 52 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-KWy7WK_DRBF3g9y-gZl:-KXClM9KWmlehZS9qJCJ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/849961874e810bcb94acb586a458e576b23084d7/drake/systems/controllers/gravity_compensator.cc#L52)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, an assert shouldn't ask for mutable access which might trigger computation.
</details>
Done.

---

*[drake/systems/controllers/gravity_compensator.cc, line 21 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-KXD8B76-0K7jS3S57MZ:-KXDoWQedj2rlueA5PyH:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5dee5c4eb01760f728c7cd8cdf415616a676906e/drake/systems/controllers/gravity_compensator.cc#L21)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> ` DRAKE_ABORT_MSG` for printing messages consistently. I think at this point (with all the utilities in `drake_assert.h`) we should be avoiding explicit calls to `throw`?
</details>
Done.

---

*[drake/systems/controllers/test/gravity_compensator_test.cc, line 28 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-KX5fQu_pTeoXh20zBeK:-KXCwHn8iSgZlQ9t3wZb:bf9sf2s) ([raw file](https://github.com/robotlocomotion/drake/blob/849961874e810bcb94acb586a458e576b23084d7/drake/systems/controllers/test/gravity_compensator_test.cc#L28)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary>
> This method might need renaming to ComputeRobotGravityTorque since it is being used for the 4 bar linkage example below as well.
</details>
Done. I re-named it to be `ComputeGravityTorque()`.

---

*[drake/systems/controllers/test/gravity_compensator_test.cc, line 41 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-KWy82hQGeXYZh45kPUz:-KXCl_Lh_9kEnjwRJ_ly:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/849961874e810bcb94acb586a458e576b23084d7/drake/systems/controllers/test/gravity_compensator_test.cc#L41)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, this code duplicated from gravity_compensator -- can you avoid that?
</details>
Done.

---

*[drake/systems/controllers/test/gravity_compensator_test.cc, line 50 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-KX5cbnpRXdQptzEroBP:-KXCxAJlPbZjCm-Sye7G:bjfor22) ([raw file](https://github.com/robotlocomotion/drake/blob/849961874e810bcb94acb586a458e576b23084d7/drake/systems/controllers/test/gravity_compensator_test.cc#L50)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary>
> BTW: as an alternative to repeating the computation inside gravity compensation, Im wondering if its better to compute the gravity torque, feed it into the system being controlled and then demonstrate that the resulting motion is of 0 acceleration and 0 velocity instead. This might overcome hidden problems in repeating the controller logic as a verification. 
</details>
Interesting idea. I think feeding in the computed gravity torque into a simulator may push the limits on what should be considered a "unit test". Perhaps an alternative is to manually derive the correct value and simply check that it matches the output of  the `GravityCompensator`. Let's push this into a future PR: #4231.

---

*[drake/systems/controllers/test/gravity_compensator_test.cc, line 106 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-KWy8gSzP0vvbrDxnHoK:-KXCxqm-fvkDwTjJTXmA:b-ojelbs) ([raw file](https://github.com/robotlocomotion/drake/blob/849961874e810bcb94acb586a458e576b23084d7/drake/systems/controllers/test/gravity_compensator_test.cc#L106)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Yes, the subsequent code should not (and does not) access `tree`, but it does access `tree_`. I've removed the comment but kept the scoping curly brace just to reinforce the face that `tree_` is the only choice.
</details>
I removed the scoping.

---

*[drake/systems/controllers/test/gravity_compensator_test.cc, line 142 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-KWy9F5uW1c0I2XAI9yo:-KXCxtKQzGu21LXRtKsA:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/849961874e810bcb94acb586a458e576b23084d7/drake/systems/controllers/test/gravity_compensator_test.cc#L142)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, nuke most of this comment?
</details>
Done.

---

*[drake/systems/controllers/test/gravity_compensator_test.cc, line 48 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-KXD7dTh0MWRuPoAPVIN:-KXDom6xGbj1iMGd9r0h:bwxduxq) ([raw file](https://github.com/robotlocomotion/drake/blob/5dee5c4eb01760f728c7cd8cdf415616a676906e/drake/systems/controllers/test/gravity_compensator_test.cc#L48)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> do we need the empty `SetUp()`?
</details>
Yes, without this, the following compile-time error occurs:

```
/Users/liang/dev/drake-distro/build/install/include/gtest/gtest.h:430:16: note: hidden overloaded virtual function 'testing::Test::SetUp' declared here: different number of parameters (0 vs 1)
  virtual void SetUp();
               ^
1 error generated.
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-:-KXDpO3tkW70JoGeresk:bblnx7v)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 7 of 10 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/controllers/gravity_compensator.cc, line 21 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-KXD8B76-0K7jS3S57MZ:-KXGWbjGBHB6OOex2KX_:b-wyxqc5) ([raw file](https://github.com/robotlocomotion/drake/blob/5dee5c4eb01760f728c7cd8cdf415616a676906e/drake/systems/controllers/gravity_compensator.cc#L21)):*
> ERROR: GravityCompensator: 

Similar message printed by `DRAKE_ABORT_MSG`. `msg` should only contain the actual message, no location, no file, no "ERROR" or similar.

---

*[drake/systems/controllers/test/gravity_compensator_test.cc, line 48 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-KXD7dTh0MWRuPoAPVIN:-KXGXHeWcAVRhMtYCj7u:bweqhzc) ([raw file](https://github.com/robotlocomotion/drake/blob/5dee5c4eb01760f728c7cd8cdf415616a676906e/drake/systems/controllers/test/gravity_compensator_test.cc#L48)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Yes, without this, the following compile-time error occurs:
> 
> ```
> /Users/liang/dev/drake-distro/build/install/include/gtest/gtest.h:430:16: note: hidden overloaded virtual function 'testing::Test::SetUp' declared here: different number of parameters (0 vs 1)
>   virtual void SetUp();
>                ^
> 1 error generated.
> ```
</details>
I see. Probably then it is a bad idea to "overload" `SetUp` as you do below with exactly the same name as in the gtest? Maybe just rename `SetUp(unique_ptr<RBT> tree)` bellow to something else? I guess I would be ok with it (gets the job done) but not sure if it is "best practice".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4099)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 7 of 10 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-:-KXGaq9nQSGAUf1IT3Nk:bnfp4nl)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@amcastro-tri I addressed your comments. PTAL.

---

Review status: 5 of 10 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/controllers/gravity_compensator.cc, line 21 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-KXD8B76-0K7jS3S57MZ:-KXGblTFXN0zXLa4c_9S:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5dee5c4eb01760f728c7cd8cdf415616a676906e/drake/systems/controllers/gravity_compensator.cc#L21)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> > ERROR: GravityCompensator: 
> 
> Similar message printed by `DRAKE_ABORT_MSG`. `msg` should only contain the actual message, no location, no file, no "ERROR" or similar.
</details>
Done.

---

*[drake/systems/controllers/test/gravity_compensator_test.cc, line 48 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-KXD7dTh0MWRuPoAPVIN:-KXGbmO0aIatTIlFnmxO:b-p3s3cj) ([raw file](https://github.com/robotlocomotion/drake/blob/5dee5c4eb01760f728c7cd8cdf415616a676906e/drake/systems/controllers/test/gravity_compensator_test.cc#L48)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> I see. Probably then it is a bad idea to "overload" `SetUp` as you do below with exactly the same name as in the gtest? Maybe just rename `SetUp(unique_ptr<RBT> tree)` bellow to something else? I guess I would be ok with it (gets the job done) but not sure if it is "best practice".
</details>
Done. I pushed most of the code into the test fixture class.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-:-KXGc6qK56W5avBU_MUp:bqp3ykv)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 5 files at r5.
Review status: 9 of 10 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-:-KXGmQWROFnoyaBoCM0b:bnfp4nl)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Reviewed 1 of 5 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4099)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Thanks. Now waiting on CI to complete.

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4099#-:-KXGnMGRq1U0PdDRhhh1:b3pc2w6)*
<!-- Sent from Reviewable.io -->
